### PR TITLE
Fix an error on SSL config reload (plus some cleanup).

### PR DESCRIPTION
**Pull request by ywkaras**
Fix an error on SSL config reload (plus some cleanup).
    
    This seems to have eliminated some ERROR diags we were seeing in Yahoo Prod when
    doing config reloads.
    
    The SSLSecret public functions no longer return pointers into the unorded_map of
    secrets, they return a copy of the secret data. This seemed thread unsafe. A
    periodic poll running in the background can update the secret data for an entry
    for a secret name in the map.
    
    To avoid exporting pointers, I had to change the prototype of TSSslSecretGet().
    Hopefully there are no existing plugins that are already using this TS API function,
    so breaking this rule will be moot.

### CODE REVIEW

**Comment by ywkaras** [comment]
This PR replaces https://github.com/apache/trafficserver/pull/8790 , which was laid waste by zwoop's merciless purge of personal branches in the main repo.

The TSSslSecretGet() function is now compatible with the rest of the TS API (and bell bottoms and stagflation).

File: iocore/net/SSLUtils.cc [hunk 1]
```diff
@@ -1053,8 +1053,8 @@ SSLPrivateKeyHandler(SSL_CTX *ctx, const SSLConfigParams *params, const char *ke
     scoped_BIO bio(BIO_new_mem_buf(secret_data, secret_data_len));
     pkey = PEM_read_bio_PrivateKey(bio.get(), nullptr, nullptr, nullptr);
     if (nullptr == pkey) {
-      Debug("ssl_load", "failed to load server private key from %s",
-            (!keyPath || keyPath[0] == '\0') ? "[empty key path]" : keyPath);
+      SSLError("failed to load server private key (%.*s) from %s", secret_data_len < 50 ? secret_data_len : 50, secret_data,
```
Comment (maskit) [inline iocore/net/SSLUtils.cc:1056]: I'm not sure when to use SSLError instead of Debug, but do the battle on another PR.

**Comment by maskit** [comment]
Oh no, we released 9.2.0 without this fix. This means 9.2.0 has a problematic TS API and we can't fix it until 10.0.

**Comment by maskit** [comment]
@zwoop What do you think about releasing 9.2.1 soon with this technically an incompatible change?

File: iocore/net/SSLUtils.cc (hunk 1 above)
Comment (ywkaras) [reply iocore/net/SSLUtils.cc:1056]: Errors go into diags.log, so it's things that the operations people should see.

Comment (maskit) [reply iocore/net/SSLUtils.cc:1056]: I'm not joining the battle. I'm saying focus on fixing the API issue on this PR.

Comment (ywkaras) [reply iocore/net/SSLUtils.cc:1056]: Yahoo is not currently using TSSslSecretGet().  For us, the main purpose of this change was to fix ERRORs in diags.log that we were seeing on config reload.  The change to the API was just part of the effort to prevent future problems as well as the one we were seeing.  Many problems happen first (or only) in prod, and will go unnoticed if they are only seen in debug trace rather than in diags.log.

Comment (maskit) [reply iocore/net/SSLUtils.cc:1056]: I don't see a reason to do the two things at the same time on one PR.

This line was changed to Debug about two months ago for some reasons. Changing it back to SSLError would need a discussion. I don't want to keep the API problematic until the discussion ends.

Comment (ywkaras) [reply iocore/net/SSLUtils.cc:1056]: Do you have a link to that discussion?  If bad configuration files are not an error, what is?'

This is a cleanup PR, so it does many, many things at once.  If you want me to break it out into a large number of small PRs, I'll have to do that in the future as a low-priority task.  Otherwise, Yahoo would fire me for having very very poor judgement about priorities ha ha.

Comment (maskit) [reply iocore/net/SSLUtils.cc:1056]: The discussion is not started yet. You should start the discussion if you want to change this line back to SSLError. I don't think you should rewrite it as you want without understanding why it was changed to Debug. And the previous change was approved by you. #9225. What was it?

> This is a cleanup PR

I don't understand why you still call this just a clean up. You mean the API is fine as is?

Comment (ywkaras) [reply iocore/net/SSLUtils.cc:1056]: @serrislew why did you change this to a debug trace?

Comment (ywkaras) [reply iocore/net/SSLUtils.cc:1056]: It's a cleanup PR because it includes many relatively small changes.  One was fixing a potential bug if the TSSslSecretGet() API call was used.  The other changes fixed the config reload error Yahoo was seeing, but I can't identify which change or set of changes constituted the fix.

Comment (serrislew) [reply iocore/net/SSLUtils.cc:1056]: It still calls a SSLError log https://github.com/apache/trafficserver/pull/9225/files#diff-cb865c0bc65fb8ef103a206282b78a50e0c1c9e93ca6713322ba9df9d921e6c9L2453

Comment (ywkaras) [reply iocore/net/SSLUtils.cc:1056]: But that's a different line of code.

Comment (serrislew) [reply iocore/net/SSLUtils.cc:1056]: But that additional SSLError log replaces all of the changed SSLError to Debug calls. `SSLPrivateKeyHandler` is only called once and it will reach that additional SSLError when it returns false. And now with the SSLError called after `SSLPrivateKeyHandler` return false, it can print the cert name that wasn't available inside the function

Comment (ywkaras) [reply iocore/net/SSLUtils.cc:1056]: Thanks Serris.

File: tests/gold_tests/tls/tls_client_cert2_plugin.test.py [hunk 2]
```diff
@@ -105,7 +105,7 @@
     '  client_cert: {0}/../signed-bar.pem'.format(ts.Variables.SSLDir),
     '  client_key: {0}/../signed-bar.key'.format(ts.Variables.SSLDir),
     '- fqdn: bob.*.com',
-    '  client_cert: {0}/../combo-signed-foo.pem'.format(ts.Variables.SSLDir),
+    '  client_cert: {0}/combo-signed-foo.pem'.format(ts.Variables.SSLDir),
```
Comment (maskit) [inline tests/gold_tests/tls/tls_client_cert2_plugin.test.py:108]: This file is in the same directory as other files. Why just this one?

Comment (ywkaras) [reply tests/gold_tests/tls/tls_client_cert2_plugin.test.py:108]: I don't know.  I made that change and the test passed.

Comment (ywkaras) [reply tests/gold_tests/tls/tls_client_cert2_plugin.test.py:108]: It seems that combo-signed-foo.pem is the only cert or key listed in sni.yaml that is actually needed:
```
wkaras ~/REPOS/YTS/tests/_sandbox/tls_client_cert2_plugin/ts
O$ ls
bin  cache  config  log  plugin  runtime  ssl  storage
wkaras ~/REPOS/YTS/tests/_sandbox/tls_client_cert2_plugin/ts
O$ ls ssl
combo-signed-foo.pem  server.pem       signed2-foo.pem  signed-bar.pem  signed-foo.pem
server.key            signed2-bar.pem  signed-bar.key   signed-foo.key
wkaras ~/REPOS/YTS/tests/_sandbox/tls_client_cert2_plugin/ts
O$ cat config/sni.yaml 
sni:
- fqdn: bob.bar.com
  client_cert: /home/wkaras/REPOS/YTS/tests/_sandbox/tls_client_cert2_plugin/ts/ssl/../signed-bar.pem
  client_key: /home/wkaras/REPOS/YTS/tests/_sandbox/tls_client_cert2_plugin/ts/ssl/../signed-bar.key
- fqdn: bob.*.com
  client_cert: /home/wkaras/REPOS/YTS/tests/_sandbox/tls_client_cert2_plugin/ts/ssl/combo-signed-foo.pem
- fqdn: "*bar.com"
  client_cert: /home/wkaras/REPOS/YTS/tests/_sandbox/tls_client_cert2_plugin/ts/ssl/../signed2-bar.pem
  client_key: /home/wkaras/REPOS/YTS/tests/_sandbox/tls_client_cert2_plugin/ts/ssl/../signed-bar.key
- fqdn: "foo.com"
  client_cert: /home/wkaras/REPOS/YTS/tests/_sandbox/tls_client_cert2_plugin/ts/ssl/../signed2-foo.pem
  client_key: /home/wkaras/REPOS/YTS/tests/_sandbox/tls_client_cert2_plugin/ts/ssl/../signed-foo.key
wkaras ~/REPOS/YTS/tests/_sandbox/tls_client_cert2_plugin/ts
O$
```

Comment (ywkaras) [reply tests/gold_tests/tls/tls_client_cert2_plugin.test.py:108]: OK, when I remove my change, the test still works.  I don't remember why I made it.

**Review by maskit — APPROVED** [review]

**Comment by zwoop** [comment]
Is this something we want for 9.2.1 ?

**Comment by ywkaras** [comment]
I would recommend backporting, I'll do if you want, in case there are any conflicts.

**Comment by ywkaras** [comment]
It will change the prototype of one TS API function.  But the current prototype for that function (which returns a pointer to data that's modified by multiple threads) is inherently crashy to use (hopefully no one's using it).

**Comment by maskit** [comment]
We talked about it on weekly bug scrub. The API was newly added on 9.2.0. It's now (before many people start using it) or 10.0. Fixing it on 9.2.2 doesn't make sense.
